platforms/software/maven/src/main/java/org/gradle/api/publish/maven/internal/artifact/AbstractMavenArtifact.java

```diff
@@ -28,43 +31,42 @@
 public abstract class AbstractMavenArtifact implements MavenArtifact, PublicationArtifactInternal {
     private final TaskDependency allBuildDependencies;
     private final DefaultTaskDependency additionalBuildDependencies;
-    private String extension;
-    private String classifier;
-
-    protected AbstractMavenArtifact(TaskDependencyFactory taskDependencyFactory) {
+    private final Property<String> extensionProperty;
+    private final Property<String> classifierProperty;
+
+    protected AbstractMavenArtifact(
+        TaskDependencyFactory taskDependencyFactory,
+        ObjectFactory objectFactory,
+        ProviderFactory providerFactory
+    ) {
         this.additionalBuildDependencies = new DefaultTaskDependency();
         this.allBuildDependencies = taskDependencyFactory.visitingDependencies(context -> {
             context.add(getDefaultBuildDependencies());
             additionalBuildDependencies.visitDependencies(context);
         });
+        this.extensionProperty = objectFactory.property(String.class);
+        this.classifierProperty = objectFactory.property(String.class);
+        // those should be lazy because the fields are not yet initialized in child classes
+        getExtension().set(providerFactory.provider(() -> getDefaultExtension().getOrNull()));
+        getClassifier().set(providerFactory.provider(() -> getDefaultClassifier().getOrNull()));
     }
 
     @Override
     public abstract File getFile();
 
     @Override
-    public final String getExtension() {
-        return extension != null ? extension : getDefaultExtension();
+    public Property<String> getExtension() {
```

Comment:
❓ Wy cant these be abstract?

Reply:
Unfortunately, no, since we create artifacts with a constructor